### DISCUSSION TITLE
Receiving data with constant time step and timestamp

### DISCUSSION THREAD

**Discussion by TE-2** [FreeOpcUa/opcua-asyncio#581] [Q&A]
Hello everyone!

I am working on an OPC UA client using the FreeOPCUA library to regularly receive drive data with timestamp at a defined time rate from a PLC.
Using UAExpert I have already managed to retrieve the data with a defined time interval by using the DataChangeFilter settings ("Status/Value/Timestamp") and Monitoring Mode: Reporting.

![Unbenannt](https://user-images.githubusercontent.com/84192658/119821073-6fae7780-bef2-11eb-9ef2-9913e9d6a05d.JPG)

here even with a sampling interval of 52ms via DataLogger:
![Unbenannt2](https://user-images.githubusercontent.com/84192658/119823771-7b4f6d80-bef5-11eb-8766-b941605de8aa.JPG)

However, I fail to implement these settings with the FreeOPCUA library.
I have managed to create a subscribe to a variable but neither a constant rate nor a timestamp to receive:

```
import sys
sys.path.insert(0, "..")
from asyncua.sync import Client


class SubHandler(object):

    def datachange_notification(self, node, val, data):
        print("Python: New data change event", node, val)

    def event_notification(self, event):
        print("Python: New event", event)

try:
    client = Client('opc.tcp://' + IP_Adresse + ':' + Port + '/freeopcua/server/')
    print("Connecting to: " 'opc.tcp://' + IP_Adresse + ':' + Port + '/freeopcua/server/')
    client.connect()

except:
    exit_routine("Connection has failed")

handler = SubHandler()
Subscription_mynode = client.create_subscription(200, handler)
handle = Subscription_mynode.subscribe_data_change(mynode)

etc.
```

The next thought was to use the Subscription.create_monitored_items() method, which I  reconsider after [this forum entry](https://github.com/FreeOpcUa/python-opcua/issues/469)

Is this the right way to implement my requirements or can I adjust a timestamp and the DataChangeFilter in the client.create_subscription method somehow?

I would be very thankful for any support :)

**Comment by oroulet** [comment]
I am not sure what that timestamp option is about, I do not see it in any parameters, but it looks like you can create your subscribotion then just call sub.modify_monitored_item to chane queue size and deadband

**Reply by TE-2** [reply]
Thank you for the quick answer that helps me already further!

to be able to use the modify_monitored_item method I have to use the Subscription class and instantiate it beforehand.
is this the correct way?:
```
...
from asyncua.common.subscription import Subscription
...

handler = SubHandler()
Subscription_mynode = client.create_subscription(200, handler)

subscription = Subscription(server = UAClient)
subscription.modify_monitored_item(new_samp_time = 200, new_queuesize = 1, mod_filter_val =0)

handle = Subscription_PositionvMaster.subscribe_data_change(mynode) 
```
I am not sure if this is correct, because somewhere I would have to set the reference to the variable(mynode).

Could you please give me an example which parameters I have to pass to the constructor so that the call of the method refers to the subscription of my variable. That would be very helpful!

I have not been working with the library for very long and I am still trying to understand how it works.

**Comment by AndreasHeine** [comment]
If you need data in a fixed interval you need to poll it!
the subscription is not ment to be used for that, a subscription works like a good old mailbox which got emptied in a fixed interval and each monitored item puts a letter (datachangenotification) into the mailbox everytime the value changes (which is not automatically in a fixed interval!)

even if you use sampling instead of reporting it could be sometimes different.

edit:

if you really want to use the subscription for that you can resample the data to a fixed timeinterval with pandas!?

**Reply by TE-2** [reply]
yes ok you are probably right. I think you mean this issue:

![Unbenannt](https://user-images.githubusercontent.com/84192658/119852528-a9da4200-bf0f-11eb-9586-f1dff58b453e.JPG)

I thought only if i have the source timestamp i can track when exactly the file was written.

with Poll do you mean to read the value of the variable regularly via get_value() or is there an efficient service for this in the OPCUA architecture?

**Reply by AndreasHeine** [reply]
if you need to poll many nodes you can user read_values() instead of read_value()
```python

   print("-----------------------------------------------------")
    # read one node attribute at a time
    node = client.get_node("ns=0;i=2267") # ServiceLevel
    print("Value: ", await node.read_value())
    print("DataValue: ", await node.read_data_value())
    print("BrowseName: ", await node.read_browse_name())
    print("DataType: ", await node.read_data_type())
    print("DisplayName: ", await node.read_display_name())

    print("-----------------------------------------------------")
    # read multiple node values at once
    print("ReadValues:")
    result = await client.read_values([
        client.get_node("ns=0;i=2267"), 
        client.get_node("ns=0;i=2256")
        # ...
    ])
    for res in result:
        print(res)

    print("-----------------------------------------------------")  
    # read multiple node attributes at once
    print("ReadAttributeValues:")
    result = await client.uaclient.read_attributes([
        client.get_node("ns=0;i=2267").nodeid, 
        client.get_node("ns=0;i=2256").nodeid
        # ...
        ], 
        ua.AttributeIds.Value
    )
    for res in result:
        print(res)


```

**Reply by AndreasHeine** [reply]
the last one is if you want to read other attributes as value (it actually does the same as read_values() in the background but you can specify the attributeid e.g. BrowseName, ...)

**Reply by TE-2** [reply]
so for example if i want to have data from 2 of my nodes every 200ms i could just run the read_values() function every 200ms ?

```python
sampling_rate = 200e-3  # 200 ms sleep

async def periodic():
    print("ReadValues:")
    while True:
        result = await client.read_values([
        client.get_node("ns=0;i=2267"), 
        client.get_node("ns=0;i=2256")])

        for res in result:
            print(res)
            
        await asyncio.sleep(sampling_rate)


def stop():
    task.cancel()

loop = asyncio.get_event_loop()
loop.call_later(5, stop)            #führe 5 sekunden aus
task = loop.create_task(periodic())

try:
    loop.run_until_complete(task)
except asyncio.CancelledError:
    pass
```

Since requesting and receiving the values also takes time how can I be sure that the data actually has the desired time interval?
Is there a way to request the source timestamp for read_values() so that I can see exactly when the data was read ?

Thanks for the effort!

**Reply by AndreasHeine** [reply]
no not with read value... propably with historizing opc ua server but why not sleep for a calculated time not a fixed interval like interval minus request/response time?

**Reply by TE-2** [reply]
okay so you mean something like this:

```python
sampling_rate = 200e-3  # 200 ms sleep

async def periodic():
    print("ReadValues:")
    while True:
        start = time.time()
        result = await client.uaclient.read_attributes([client.get_node("ns=0;i=2267").nodeid, client.get_node("ns=0;i=2256").nodeid],ua.AttributeIds.Value)

        for res in result:
            print(res)
        
        end = time.time()
        time_delay = end - start
        await asyncio.sleep(sampling_rate - time_delay)


def stop():
    task.cancel()

loop = asyncio.get_event_loop()
loop.call_later(5, stop)            #führe 5 sekunden aus
task = loop.create_task(periodic())

try:
    loop.run_until_complete(task)
except asyncio.CancelledError:
    pass
```

but is there any way to get and print the timestamps of the read values only for validation ?

**Reply by mahadi** [reply]
@AndreasHeine: As i have a smilar use case i wonder if the polling performance is worse than using a subscription? I know this is a very general question, but is there any advice from your side you can give?

**Reply by AndreasHeine** [reply]
@mahadi the thing is with a subscriptions and monitoreditems you only transport changes... not data at a const timerate!

> https://reference.opcfoundation.org/Core/Part4/v104/docs/5.12.1.2
> The assigned sampling interval defines a “best effort” cyclic rate that the [Server](https://reference.opcfoundation.org/search/17?t=Server) uses to sample the item from its source. “Best effort” in this context means that the [Server](https://reference.opcfoundation.org/search/17?t=Server) does its best to sample at this rate. Sampling at rates faster than this rate is acceptable, but not necessary to meet the needs of the [Client](https://reference.opcfoundation.org/search/17?t=Client). How the [Server](https://reference.opcfoundation.org/search/17?t=Server) deals with the sampling rate and how often it actually polls its data source internally is a [Server](https://reference.opcfoundation.org/search/17?t=Server) implementation detail.

sampling in the context is how often the server checks the datasource of a node for changes and according to filters e.g. deadband it creates a notification and hand it over to the subscription which  handles the transport to the client in the next publish request/response.

what i did in the past is to resample the data recv. by the client (e.g. with pandas) if i need a const time between values e.g. for a plot

**Reply by AndreasHeine** [reply]
even with polling you can have a failing call/request and the time between calls is not const plus on top of that network calls via tcp are non deterministic as well!
in critical usecases you need to have HistoryAccess on the opc ua server its currently the most resilient solution, so even if the connection breaks and all the recovery of session and subscription fails HistoryAccess still works without dataloss

**Reply by mahadi** [reply]
@AndreasHeine: Yes, this is actually my main concern. So i understand that a server will typically not be able to manage different sampling rates for different monitored items (given a sampling interval greater than its minimum sampling interval). I guess this would be difficult to implement.
Using the SourceTimeStamp which cotains the sampling time on the server a resampling could be performed on the client side. But is this SourceTimeStamp available always? In the examples i see that it is None.

**Reply by AndreasHeine** [reply]
it depends... 

> https://reference.opcfoundation.org/Core/Part4/v104/docs/5.12.2.2
> timestampsToReturn:
> An enumeration that specifies the timestamp [Attributes](https://reference.opcfoundation.org/search/17?t=Attributes) to be transmitted for each [MonitoredItem](https://reference.opcfoundation.org/search/17?t=MonitoredItem). The [TimestampsToReturn](https://reference.opcfoundation.org/search/17?t=TimestampsToReturn) enumeration is defined in [7.35](https://reference.opcfoundation.org/Core/Part4/v104/docs/?r=_Ref133134319).
> 
> When monitoring [Events](https://reference.opcfoundation.org/search/17?t=Events), this applies only to [Event ](https://reference.opcfoundation.org/search/17?t=Event)fields that are of type [DataValue](https://reference.opcfoundation.org/search/17?t=DataValue).

> 
> https://reference.opcfoundation.org/Core/Part4/v104/docs/7.35
> SOURCE_0 | Return the source timestamp.
> SERVER_1 | Return the Server timestamp.
> BOTH_2 | Return both the source and Server timestamps.
> NEITHER_3 | Return neither timestamp.                         This is the default value for MonitoredItems if a Variable value is not being accessed.                         For HistoryRead this is not a valid setting.
> INVALID_4 | No value specified.

**Reply by AndreasHeine** [reply]
if you use Both in the CreateMonitoredItems-Request it should return the timestamps!
the thing with implementation is that you usually set a nodes value with a DataValue which contains both timestamps. however i have seen some implementations where on timestamp is missing or set incorrectly... but thats not the normal case!

**Reply by mahadi** [reply]
Ok, sounds good, then i just need to make up my mind how i can handle differences in the PLC clock and my client clock. They are not guaranteed to be synchronised. Thanks a lot for the information and explanations!

**Reply by AndreasHeine** [reply]
opc ua by spec has UTC time! but changing it to local time is simple -> [datetime -> tzinfo](https://docs.python.org/3/library/datetime.html#datetime.datetime.astimezone)
